### PR TITLE
Add Redis Streams

### PR DESCRIPTION
**Pull request by gkorland**
This is an initial PR for Redis Streams, covering most the API.

### CODE REVIEW

**Comment by gkorland** [comment]
ref #1820

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 1]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
     return client.getIntegerReply();
   }
 
+  @Override
+  public EntryID xadd(final String key, final EntryID id, final Map<String, String> hash) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Is it possible to have a `EntryID.asterisk()` for this purpose?

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 2]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+    return xadd(key, id, hash, Long.MAX_VALUE, true);
+  }
+  
+  @Override
+  public EntryID xadd(final String key, EntryID id, final Map<String, String> hash, final long maxLen, final boolean exactMaxLen) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: How about a `xadd(final String key, EntryID id, final Map<String, String> hash, final long maxLen)` version?

Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: In Redis, extra `~` means approximate length. So, IMHO, `approximateLength` is better as parameter than `exactMaxLen`.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 3]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+
+  @Override
+  public Long xlen(final String key) {
+    checkIsInMultiOrPipeline();
+    client.xlen(key);
+    return client.getIntegerReply();
+  }
+
+  @Override
+  public List<StreamEntry> xrange(final String key, final EntryID start, final EntryID end, final int count) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: A `xrange(key, start, end)` version?

Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Is there any option to send `-` and `+` for start and end?

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 4]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+    checkIsInMultiOrPipeline();
+    client.xrange(key, start, end, count);
+    return BuilderFactory.STREAM_ENTRY_LIST.build(client.getObjectMultiBulkReply());
+  }
+  
+  @Override
+  public List<StreamEntry> xrevrange(final String key, final EntryID end, final EntryID start, final int count) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Same as `xrange`.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 5]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+    } finally {
+      client.rollbackTimeout();
+    }
+  }
+
+  @Override
+  public long xack(final String key, final String group, final EntryID... ids) {
+    checkIsInMultiOrPipeline();
+    client.xack(key, group, ids);
+    return client.getIntegerReply().longValue();
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Why `.longValue()`? Shouldn't it be auto-casted?

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 6]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+  }
+
+  @Override
+  public String xgroupCreate(final String key, final String groupname, final EntryID id, final boolean makeStream) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: A `xgroupCreate(key, groupname, id)` version?

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 7]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+
+  @Override
+  public long xdel(final String key, final EntryID... ids) {
+    checkIsInMultiOrPipeline();
+    client.xdel(key, ids);
+    return client.getIntegerReply();
+  }
+
+  @Override
+  public long xtrim(final String key, final long maxLen, final boolean exactMaxLen) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Same as `xadd`.

File: src/main/java/redis/clients/jedis/EntryID.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
+package redis.clients.jedis;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+public class EntryID implements Comparable<EntryID>, Serializable{
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/EntryID.java]: How about `StreamEntryID` instead of `EntryID`?

File: src/main/java/redis/clients/jedis/PendingEntry.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
+package redis.clients.jedis;
+
+import java.io.IOException;
+import java.io.Serializable;
+
+public class PendingEntry implements Serializable{
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/PendingEntry.java]: How about `StreamPendingEntry` instead of `PendingEntry`?

File: src/main/java/redis/clients/jedis/MultiKeyPipelineBase.java [hunk 10]
```diff
@@ -701,6 +701,5 @@
       final int timeout, final MigrateParams params, final byte[]... keys) {
     client.migrate(host, port, destinationDB, timeout, params, keys);
     return getResponse(BuilderFactory.STRING);
-  }
-  
+  }
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/MultiKeyPipelineBase.java]: Unrelated change.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 11]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+    client.xpending(key, groupname, start, end, count, consumername);
+
+    // TODO handle consumername == NULL case
+    
+    return BuilderFactory.PENDING_ENTRY_LIST.build(client.getObjectMultiBulkReply());
+  }
+
+  @Override
+  public List<StreamEntry> xclaim(String key, String group, String consumername, long minIdleTime, long newIdleTime,
+      int retries, boolean force, EntryID... ids) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Please change to `Integer retries`, `Boolean force`, etc. so that user can send `null` when not sure.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 12]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+    checkIsInMultiOrPipeline();
+    client.xpending(key, groupname, start, end, count, consumername);
+
+    // TODO handle consumername == NULL case
+    
+    return BuilderFactory.PENDING_ENTRY_LIST.build(client.getObjectMultiBulkReply());
+  }
+
+  @Override
+  public List<StreamEntry> xclaim(String key, String group, String consumername, long minIdleTime, long newIdleTime,
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: A version with only compulsory params? E.g. `xclaim(key, group, consumername, minIdleTime, ids)`.

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 11 above)
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: A version with `Params` can be used here. Skip it for later time if inconvenient.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 13]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+   * @param key
+   * @param groupname
+   * @param cosumer
+   * @param count
+   * @param block
+   * @param streams
+   * @return
+   */
+  @Override
+  public List<Entry<String, List<StreamEntry>>> xreadGroup(final String groupname, final String consumer, final int count, final long block,
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Why not use `Map` here?

Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Same as `xclaim`.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 14]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
...
+      List<Object> stream = (List<Object>)streamObj;
+      String streamId = SafeEncoder.encode((byte[])stream.get(0));
+      List<StreamEntry> streamEntries = BuilderFactory.STREAM_ENTRY_LIST.build(stream.get(1));
+      result.add(new AbstractMap.SimpleEntry<String, List<StreamEntry>>(streamId, streamEntries));
+    }
+    return result;
+  }
+
+  @Override
+  public List<PendingEntry> xpending(final String key, final String groupname, final EntryID start, final EntryID end,
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: Same as `xclaim`.

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: I don't users should deal directly with String based Ids, the streams expect the id to be in a specific format

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 15]
```diff
@@ -30,7 +30,7 @@
 import redis.clients.jedis.util.Slowlog;
 
 public class Jedis extends BinaryJedis implements JedisCommands, MultiKeyCommands,
-    AdvancedJedisCommands, ScriptingCommands, BasicCommands, ClusterCommands, SentinelCommands, ModuleCommands {
+AdvancedJedisCommands, ScriptingCommands, BasicCommands, ClusterCommands, SentinelCommands, ModuleCommands {
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: done

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 16]
```diff
@@ -3645,4 +3645,175 @@ public Long hstrlen(final String key, final String field) {
     return client.getIntegerReply();
   }
 
+  @Override
+  public EntryID xadd(final String key, final EntryID id, final Map<String, String> hash) {
+    return xadd(key, id, hash, Long.MAX_VALUE, true);
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: In case on MAX_VALUE the MAXLEN is not serialized on the command call. see: https://github.com/gkorland/jedis/blob/d7fceb249ec1982d7727a4597de427da052c4356/src/main/java/redis/clients/jedis/BinaryClient.java#L1263

If the user wants to set the MAXLEN he can call the overloaded xadd: https://github.com/gkorland/jedis/blob/d7fceb249ec1982d7727a4597de427da052c4356/src/main/java/redis/clients/jedis/Jedis.java#L3654

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 1 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: Added EntryID.NEW_ENTRY

We have similar for other APIs like:
EntryID.LAST_ENTRY --> "$"
EntryID.UNRECEIVED_ENTRY --> ">"

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 2 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: I didn't want to overload the too many methods, so I created the a simple and a full.

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 5 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: done, thanks

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 6 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: I wasn't sure what should be the default behavior for makeStream, so I left it like that for now

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 7 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: my feeling is that the default expected behavior is exactMaxLen=true therefor I think it should be exactMaxLen

File: src/main/java/redis/clients/jedis/EntryID.java (hunk 8 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/EntryID.java]: Since there's no real association to a specific Stream and the same EntryID can be used with couple of Streams I thought it's better to keep it as EntryID

File: src/main/java/redis/clients/jedis/PendingEntry.java (hunk 9 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/PendingEntry.java]: I wanted to keep it close to EntryID, do you think it's better?

File: src/main/java/redis/clients/jedis/MultiKeyPipelineBase.java (hunk 10 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/MultiKeyPipelineBase.java]: done

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 11 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: good idea, let's skip it this release

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 13 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: since there's an order

**Comment by gkorland** [comment]
@sazzad16 thanks for the great review

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 17]
```diff
@@ -3646,4 +3646,175 @@ public Long hstrlen(final String key, final String field) {
...
+
+  @Override
+  public Long xlen(final String key) {
+    checkIsInMultiOrPipeline();
+    client.xlen(key);
+    return client.getIntegerReply();
+  }
+
+  @Override
+  public List<StreamEntry> xrange(final String key, final EntryID start, final EntryID end, final int count) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: make `count` optional (Integer). sometimes we'd have to allow something even if it's risky. e.g. `keys` command

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 18]
```diff
@@ -3646,4 +3646,175 @@ public Long hstrlen(final String key, final String field) {
     return client.getIntegerReply();
   }
 
+  @Override
+  public EntryID xadd(final String key, final EntryID id, final Map<String, String> hash) {
+    return xadd(key, id, hash, Long.MAX_VALUE, true);
+  }
+  
+  @Override
+  public EntryID xadd(final String key, EntryID id, final Map<String, String> hash, final long maxLen, final boolean exactMaxLen) {
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: This is why we have the overloaded function -   
public StreamnEntryID xadd(final String key, final StreamnEntryID id, final Map<String, String> hash) {

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 17 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: a user can always send Integer.MAX_VALUE

[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: One can always pass on star or end

File: src/main/java/redis/clients/jedis/commands/MultiKeyBinaryRedisPipeline.java [hunk 19]
```diff
@@ -80,5 +83,5 @@
 
   Response<Long> touch(byte[]... keys);
 
-  Response<String> migrate(String host, int port, int destinationDB, int timeout, MigrateParams params, byte[]... keys);
+  Response<String> migrate(String host, int port, int destinationDB, int timeout, MigrateParams params, byte[]... keys);
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/MultiKeyBinaryRedisPipeline.java]: fixed

File: src/main/java/redis/clients/jedis/commands/JedisClusterCommands.java [hunk 20]
```diff
@@ -314,7 +317,168 @@
    * @param field
    * @return lenth of the value for key
    */
-  Long hstrlen(String key, String field);
+  Long hstrlen(String key, String field);
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/JedisClusterCommands.java]: fixed

File: src/main/java/redis/clients/jedis/EntryID.java [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/EntryID.java]: fixed

File: src/main/java/redis/clients/jedis/commands/MultiKeyCommandsPipeline.java [hunk 22]
```diff
@@ -1,12 +1,15 @@
 package redis.clients.jedis.commands;
 
 import redis.clients.jedis.BitOP;
+import redis.clients.jedis.EntryID;
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/MultiKeyCommandsPipeline.java]: fixed

File: src/main/java/redis/clients/jedis/commands/MultiKeyCommandsPipeline.java [hunk 23]
```diff
@@ -1,12 +1,15 @@
...
 import redis.clients.jedis.Response;
 import redis.clients.jedis.SortingParams;
+import redis.clients.jedis.StreamEntry;
 import redis.clients.jedis.ZParams;
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/MultiKeyCommandsPipeline.java]: fixed

File: src/main/java/redis/clients/jedis/commands/MultiKeyCommandsPipeline.java [hunk 24]
```diff
@@ -1,12 +1,15 @@
...
 import redis.clients.jedis.params.MigrateParams;
 
 import java.util.List;
+import java.util.Map;
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/MultiKeyCommandsPipeline.java]: fixed

File: src/main/java/redis/clients/jedis/commands/MultiKeyBinaryRedisPipeline.java [hunk 25]
```diff
@@ -1,12 +1,15 @@
 package redis.clients.jedis.commands;
 
 import redis.clients.jedis.BitOP;
+import redis.clients.jedis.EntryID;
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/MultiKeyBinaryRedisPipeline.java]: fixed

File: src/main/java/redis/clients/jedis/commands/MultiKeyBinaryRedisPipeline.java [hunk 26]
```diff
@@ -1,12 +1,15 @@
...
 import redis.clients.jedis.Response;
 import redis.clients.jedis.SortingParams;
+import redis.clients.jedis.StreamEntry;
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/MultiKeyBinaryRedisPipeline.java]: fixed

File: src/main/java/redis/clients/jedis/commands/MultiKeyBinaryRedisPipeline.java [hunk 27]
```diff
@@ -1,12 +1,15 @@
...
 import redis.clients.jedis.ZParams;
 import redis.clients.jedis.params.MigrateParams;
 
 import java.util.List;
+import java.util.Map;
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/MultiKeyBinaryRedisPipeline.java]: fixed

File: src/main/java/redis/clients/jedis/commands/Commands.java [hunk 28]
```diff
@@ -340,7 +342,7 @@ void zrevrangeByScoreWithScores(String key, String max, String min,
    * @param field
    */
   void hstrlen(String key, String field);
-
+
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/Commands.java]: fixed

File: src/main/java/redis/clients/jedis/commands/BinaryRedisPipeline.java [hunk 29]
```diff
@@ -302,7 +303,36 @@
 
   Response<List<Long>> bitfield(byte[] key, byte[]... elements);
 
-  Response<Long> hstrlen(byte[] key, byte[] field);
+  Response<Long> hstrlen(byte[] key, byte[] field);
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/commands/BinaryRedisPipeline.java]: fixed

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 30]
```diff
@@ -134,7 +136,7 @@ public String toString() {
       final Iterator<Object> iterator = flatHash.iterator();
       while (iterator.hasNext()) {
         hash.put(SafeEncoder.encode((byte[]) iterator.next()),
-          String.valueOf((Long) iterator.next()));
+            String.valueOf((Long) iterator.next()));
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/BuilderFactory.java]: fixed

File: src/main/java/redis/clients/jedis/PendingEntry.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/PendingEntry.java]: fixed

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 31]
```diff
@@ -3646,4 +3646,175 @@ public Long hstrlen(final String key, final String field) {
...
+   */
+  @Override
+  public List<Entry<String, List<StreamEntry>>> xreadGroup(final String groupname, final String consumer, final int count, final long block,
+      final boolean noAck, final Entry<String, EntryID>... streams) {
+    checkIsInMultiOrPipeline();
+    client.xreadGroup(groupname, consumer, count, block, noAck, streams);
+
+    List<Object> streamsEntries = client.getObjectMultiBulkReply();
+    if(streamsEntries == null) {
+      return new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: Don't you prefer an empty list to ease the user's coding life and avoid NPE?

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 32]
```diff
@@ -3646,4 +3646,175 @@ public Long hstrlen(final String key, final String field) {
     return client.getIntegerReply();
   }
 
+  @Override
+  public StreamnEntryID xadd(final String key, final StreamnEntryID id, final Map<String, String> hash) {
+    return xadd(key, id, hash, Long.MAX_VALUE, true);
+  }
+  
+  @Override
+  public StreamnEntryID xadd(final String key, StreamnEntryID id, final Map<String, String> hash, final long maxLen, final boolean exactMaxLen) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: IMO, default behavior should be how the command behaves without that parameter.

In Redis, extra `~` means approximate length. `approximateLength` is better parameter name than `exactMaxLen`. This way it can expect `false` by default. Moreover, isn't `false` default for:
1. xgroupCreate(boolean makeStream)
2. xreadGroup(boolean noAck)
3. xclaim(boolean force)

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 33]
```diff
@@ -3646,4 +3646,175 @@ public Long hstrlen(final String key, final String field) {
...
+
+  @Override
+  public long xdel(final String key, final StreamnEntryID... ids) {
+    checkIsInMultiOrPipeline();
+    client.xdel(key, ids);
+    return client.getIntegerReply();
+  }
+
+  @Override
+  public long xtrim(final String key, final long maxLen, final boolean exactMaxLen) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java]: IMO, default behavior should be how the command behaves without that parameter.

In Redis, extra `~` means approximate length. `approximateLength` is better parameter name than `exactMaxLen`. This way it can expect `false` by default. Moreover, isn't `false` default for:
1. xgroupCreate(boolean makeStream)
2. xreadGroup(boolean noAck)
3. xclaim(boolean force)

File: src/main/java/redis/clients/jedis/BinaryClient.java [hunk 34]
```diff
@@ -1261,4 +1261,194 @@ public void bitfield(final byte[] key, final byte[]... value) {
   public void hstrlen(final byte[] key, final byte[] field) {
     sendCommand(HSTRLEN, key, field);
   }
+  
+  public void xadd(final byte[] key, final byte[] id, final Map<byte[], byte[]> hash, long maxLen, boolean exactMaxLen) {
+      int i = 0;
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/BinaryClient.java]: fixed

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 35]
```diff
@@ -474,6 +474,87 @@ public String toString() {
 
   };
 
+  public static final Builder<StreamnEntryID> ENTRY_ID = new Builder<StreamnEntryID>() {
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/BuilderFactory.java]: Not sure I understand

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 36]
```diff
@@ -474,6 +474,87 @@ public String toString() {
...
+    public  StreamnEntryID build(Object data) {
+      if (null == data) {
+        return null;
+      }
+      return new StreamnEntryID((String)data);
+    }
+
+    @Override
+    public String toString() {
+      return "StreamEntry";
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/BuilderFactory.java]: fixed

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 37]
```diff
@@ -474,6 +474,87 @@ public String toString() {
...
+        long idleTime = BuilderFactory.LONG.build(stream.get(2));      
+        long deliveredTimes = BuilderFactory.LONG.build(stream.get(3));
+        result.add(new StreamPendingEntry(new StreamnEntryID(id), consumerName, idleTime, deliveredTimes));
+      }
+      return result;
+    }
+
+    @Override
+    public String toString() {
+      return "List<PendingEntry>";
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/BuilderFactory.java]: fixed

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 38]
```diff
@@ -3070,7 +3070,7 @@ public String sentinelFailover(final String masterName) {
   @Override
   public String sentinelMonitor(final String masterName, final String ip, final int port, final int quorum) {
     client.sentinel(Protocol.SENTINEL_MONITOR, masterName, ip, String.valueOf(port),
-      String.valueOf(quorum));
+        String.valueOf(quorum));
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: done

File: src/main/java/redis/clients/jedis/BuilderFactory.java (hunk 35 above)
[the remark this answers is not in the document]
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/BuilderFactory.java]: Rename `ENTRY_ID` to something like `STREAM_ENTRY_ID`, just for more clarity.

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 31 above)
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: done

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 32 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: notice I didn't create an overload without the "exactMaxLen", these are the two options I gave, so when a user is picking maxLen he has to consider the exactMaxLen and not assume anything about it.
xadd(key, id, hash)
xadd(key, id, hash, maxLen, exactMaxLen)

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 39]
```diff
@@ -3646,4 +3646,175 @@ public Long hstrlen(final String key, final String field) {
...
+    checkIsInMultiOrPipeline();
+    client.xpending(key, groupname, start, end, count, consumername);
+
+    // TODO handle consumername == NULL case
+    
+    return BuilderFactory.PENDING_ENTRY_LIST.build(client.getObjectMultiBulkReply());
+  }
+
+  @Override
+  public List<StreamEntry> xclaim(String key, String group, String consumername, long minIdleTime, long newIdleTime,
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: I didn't want to create too many overloads, before I'm getting feedback from users

[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: I really feel bad with boxed arguments just to let user pass null values for optional arguments

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 40]
```diff
@@ -3646,4 +3646,175 @@ public Long hstrlen(final String key, final String field) {
...
+
+  @Override
+  public Long xlen(final String key) {
+    checkIsInMultiOrPipeline();
+    client.xlen(key);
+    return client.getIntegerReply();
+  }
+
+  @Override
+  public List<StreamEntry> xrange(final String key, final StreamnEntryID start, final StreamnEntryID end, final int count) {
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: done

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 12 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: I think we should skip it for the first version and see what are using asking before we create too many overloads

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 39 above)
[the remark this answers is not in the document]
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/Jedis.java]: Well, it has to be either null values for optional arguments or too many overloads!

**Comment by gkorland** [comment]
@sazzad16 do you think we can merge and enhance on following PR?

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 41]
```diff
@@ -3646,4 +3646,170 @@ public Long hstrlen(final String key, final String field) {
...
+
+  @Override
+  public long xdel(final String key, final StreamEntryID... ids) {
+    checkIsInMultiOrPipeline();
+    client.xdel(key, ids);
+    return client.getIntegerReply();
+  }
+
+  @Override
+  public long xtrim(final String key, final long maxLen, final boolean exactMaxLen) {
```
[the remark this answers is not in the document]
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: :+1:  done.

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 33 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: :+1: done

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 32 above)
Comment (gkorland) [reply src/main/java/redis/clients/jedis/Jedis.java]: done

**Comment by huaweigu** [comment]
👍 When we will be releasing this to `3.1.0-rc*` and `3.1.0`?
